### PR TITLE
Add pass to convert scf.forall to scf.for within aie.core ops

### PR DESCRIPTION
**Pull request by jtuyls**
Adds a small utility for converting `scf.forall` operations inside `aie.core` to `scf.for`. This is a workaround for correct lowering through the AIE backend (otherwise, we get numerical errors).

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertCoreForallToFor.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    SmallVector<Operation *> forOpResults;
+    if (failed(scf::forallToForLoop(rewriter, forallOp, &forOpResults))) {
+      coreOp.emitOpError() << "failed to transform scf.forall to scf.for";
+      return WalkResult::interrupt();
+    }
+
+    SmallVector<scf::ForOp> forOps = llvm::map_to_vector(
+        forOpResults,
+        [](Operation *loop) { return dyn_cast<scf::ForOp>(loop); });
+    if (failed(coalesceLoops(rewriter, forOps))) {
```
Comment (newling) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertCoreForallToFor.cpp:39]: Please add a test with 1) perfectly nested loops and 2) not perfectly nested loops, and add comment in pass description that you also coalesce. Or don't coalesce :)

**Review by newling — CHANGES_REQUESTED** [review]

**Review by newling — APPROVED** [review]
My bad, I understand the logic now

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertCoreForallToFor.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file the conversion of `scf.forall` into `scf.for` with `aie.core`
```
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertCoreForallToFor.cpp:9]: nit: This file converts?

Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertCoreForallToFor.cpp:9]: Oopsie, had automerge on and it just merged before I clicked to disable..

Comment (nirvedhmeshram) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertCoreForallToFor.cpp:9]: No worries!

**Pull request merged**